internal/zinc-apiinfo/src/test/scala/sbt/internal/inc/ClassCanonicalNameSpec.scala

```diff
@@ -0,0 +1,118 @@
+package p4 { class x { class y { object z; class z } } }
+
+class ClassCanonicalNameSpec extends UnitSpec {
+  "ClassToAPI.classCanonicalName" should """return "" for null""" in assert(getCustomCanonicalName(null) === "")
+
+  import scala.reflect._
+  def check[T: ClassTag](expected: Expected) = checkClass(expected, classTag[T].runtimeClass, "tag")
+  def checkRef(expected: Expected, x: AnyRef) = checkClass(expected, x.getClass, "ref")
+  def checkClass(expected: Expected, c: Class[_], classSource: String) = {
+    import expected._
+    it should f"for $nesting%-5s ($classSource) return $canonicalClassName" in assert(
+      getCustomCanonicalName(c) === canonicalClassName &&
+        getNativeCanonicalName(c) === nativeCanonicalClassName
+    )
+  }
+
+  def getCustomCanonicalName(c: Class[_]) = strip(ClassToAPI.classCanonicalName(c))
+  def getNativeCanonicalName(c: Class[_]) = strip(handleMalformed(c.getCanonicalName))
+
+  // Yes, that's the only way to write these types.
+  import scala.language.existentials
+
+  check[p1.x.type](O)
+  check[p3.x](C)
+
+  check[p1.x.y.type](OO)
+  check[p2.x.y](OC)
+  check[c1.y.type forSome { val c1: p3.x }](CO)
+  check[p4.x#y](CC)
+
+  check[p1.x.y.z.type](OOO)
+  check[p1.x.y.z](OOC)
+  check[c2.z.type forSome { val c2: p2.x.y }](OCO)
+  check[p2.x.y#z](OCC)
+  check[c1.y.z.type forSome { val c1: p3.x }](COO)
+  check[c1.y.z forSome { val c1: p3.x }](COC)
+  check[c2.z.type forSome { val c2: p4.x#y }](CCO)
+  check[p4.x#y#z](CCC)
+
+  // Now again, but calling getClass on an instance instead
+  // of using the type to summon a class object. This matches
+  // against the same nestings as above.
+  checkRef(O, p1.x)
+  checkRef(C, new p3.x)
+
+  checkRef(OO, p1.x.y)
+  checkRef(OC, new p2.x.y)
+  checkRef(CO, { val c1 = new p3.x ; c1.y })
+  checkRef(CC, { val c1 = new p4.x ; new c1.y })
+
+  checkRef(OOO, p1.x.y.z)
+  checkRef(OOC, new p1.x.y.z)
+  checkRef(OCO, { val c2 = new p2.x.y ; c2.z })
+  checkRef(OCC, { val c2 = new p2.x.y ; new c2.z })
+  checkRef(COO, { val c1 = new p3.x ; c1.y.z })
+  checkRef(COC, { val c1 = new p3.x ; new c1.y.z })
+  checkRef(CCO, { val c1 = new p4.x ; val c2 = new c1.y ; c2.z })
+  checkRef(CCC, { val c1 = new p4.x ; val c2 = new c1.y ; new c2.z })
+
+  object O extends Expected("x$")
+  object C extends Expected("x")
+
+  object OO extends Expected("x.y$")
+  object OC extends Expected("x.y")
+  object CO extends Expected("x.y$")
+  object CC extends Expected("x.y")
+
+  object OOO extends Expected("x.y$$z$", nativeClassNameIsMalformed = true)
+  object OOC extends Expected("x.y$$z", nativeClassNameIsMalformed = true)
+  object OCO extends Expected("x.y.z$")
+  object OCC extends Expected("x.y.z")
+  object COO extends Expected("x.y$$z$", nativeClassNameIsMalformed = true)
+  object COC extends Expected("x.y$$z", nativeClassNameIsMalformed = true)
+  object CCO extends Expected("x.y.z$")
+  object CCC extends Expected("x.y.z")
```

Comment:
@jvican Do these canonical class names look right to you?

Looking at the pattern across all the nestings, the ones that have a malformed class name look wrong to me.

Basically I expect them all to be either "x.y.z$" or "x.y.z".

WDYT?